packages/yarnpkg-core/sources/tgzUtils.ts

```diff
@@ -69,78 +93,49 @@ export async function extractArchiveTo<T extends FakeFS<PortablePath>>(tgz: Buff
     return false;
   }
 
-  parser.on(`entry`, (header, stream, next) => {
-    if (ignore(header)) {
-      next();
-      return;
-    }
+  for await (const entry of parseTar(tgz)) {
+    if (ignore(entry))
+      continue;
 
-    const parts = ppath.normalize(npath.toPortablePath(header.name)).replace(/\/$/, ``).split(/\//g);
-    if (parts.length <= stripComponents) {
-      stream.resume();
-      next();
-      return;
-    }
+    const parts = ppath.normalize(npath.toPortablePath(entry.path)).replace(/\/$/, ``).split(/\//g);
+    if (parts.length <= stripComponents)
+      continue;
 
     const slicePath = parts.slice(stripComponents).join(`/`) as PortablePath;
     const mappedPath = ppath.join(prefixPath, slicePath);
 
     let mode = 0o644;
 
     // If a single executable bit is set, normalize so that all are
-    if (header.type === `directory` || ((header.mode ?? 0) & 0o111) !== 0)
+    if (entry.type === `Directory` || ((entry.mode ?? 0) & 0o111) !== 0)
       mode |= 0o111;
 
-    switch (header.type) {
-      case `directory`: {
+    switch (entry.type) {
+      case `Directory`: {
         targetFs.mkdirpSync(ppath.dirname(mappedPath), {chmod: 0o755, utimes: [defaultTime, defaultTime]});
 
         targetFs.mkdirSync(mappedPath);
         targetFs.chmodSync(mappedPath, mode);
         targetFs.utimesSync(mappedPath, defaultTime, defaultTime);
-        next();
       } break;
 
-      case `file`: {
+      case `OldFile`:
+      case `File`: {
         targetFs.mkdirpSync(ppath.dirname(mappedPath), {chmod: 0o755, utimes: [defaultTime, defaultTime]});
 
-        const chunks: Array<Buffer> = [];
-
-        stream.on(`data`, (chunk: Buffer) => chunks.push(chunk));
-        stream.on(`end`, () => {
-          targetFs.writeFileSync(mappedPath, Buffer.concat(chunks));
-          targetFs.chmodSync(mappedPath, mode);
-          targetFs.utimesSync(mappedPath, defaultTime, defaultTime);
-          next();
-        });
+        targetFs.writeFileSync(mappedPath, await miscUtils.bufferStream(entry as unknown as Readable));
```

Comment:
I'm not a huge fan of the double cast here because it hides the fact that the minipass and Readable types aren't entirely compatible. We could change the type of the parameter to `bufferStream` to allow for both Readables and minipass instances (e.g. by changing `Readable` into `Pick<Readable, 'on'>`).
That being said, the tests cover this code path so any problems would be caught anyway&hellip; so I'm not sure it's worth it.